DESIGN.md

```diff
@@ -1,31 +1,379 @@
+
+## Output formatting
+
+The CLI supports extensible output formatting. Formatting consists of two stages:
+
+- formatting - translating command output to a sequence of lines
+- printing - outputing the lines to some device (e.g. stdout or filesystem)
+
+A formatter module performs formatting.
+Formatter is a module, implementing the `RabbitMQ.CLI.FormatterBehaviour` behaviour:
+
+`format_output(output :: any, options :: Map.t) :: String.t | [String.t]`
+
+Format a single value, returned. It accepts output from command and named arguments (options)
+and returns a list of strings, that should be printed.
+
+`format_stream(output_stream :: Enumerable.t, options :: Map.t) :: Enumerable.t`
+
+Format a stream of return values. This function utilizes elixir
+Stream [http://elixir-lang.org/docs/stable/elixir/Stream.html] abstraction
+to define processing of continious data, so the CLI can output data in realtime.
+
+Used in `list_*` commands, that emit data in asynchronously.
+
+`DefaultOutput` will return all enumerable data as stream,
+so it will be formatted with this function.
+
+A printer module performs printing. Printer module should implement
+the `RabbitMQ.CLI.PrinterBehaviour` behaviour:
+
+`init(options :: Map.t) :: {:ok, printer_state :: any} | {:error, error :: any}`
+
+Init the internal printer state (e.g. open a file handler).
+
+`finish(printer_state :: any) :: :ok`
+
+Finalize the internal printer state
+
+`print_output(output :: String.t | [String.t], printer_state :: any) :: :ok`
+
+Print the output lines in the printer state context.
+Is called for `{:ok, val}` command output after formatting `val` using formatter,
+and for each enumerable element of `{:stream, enum}` enumerable
+
+`print_ok(printer_state :: any) :: :ok`
+
+Print an output without any values. Is called for `:ok` command return value.
+
+Output formatting logic is defined in `output.ex` module.
+
+Following rules apply for a value, returned from `output/2` callback:
+
+- `:ok` - initializes a printer, calls `print_ok` and finishes the printer. Exit code is `0`.
+- `{:ok, value}` - calls `format_output/2` from formatter, then passes the value to printer. Exit code is `0`.
+- `{:stream, enum}` - calls `format_stream/2` to augment the stream with formatting logic,
+initializes a printer, then calls `print_output/2` for each successfully formatted stream
+element (which is not `{:error, msg}`). Exit code is `0` if there were no errors.
+In case of an error element, stream processing stops, error is printed to `stderr`
+and the CLI exits with nonzero exit code.
+- `{:error, exit_code, msg}` - prints `msg` to `stderr` and exits with `exit_code`
+
+
+## Environment configuration
+
+Some commands require information about the server environment to run correctly.
+Such information is:
+
+- rabbitmq code directory
+- rabbitmq mnesia directory
+- enabled plugins file
+- plugins directory
+
+Enabled plugins and plugins directory also used to locate plugins commands.
```

Comment:
Enabled plugins and plugins directory also used to locate plugins commands => Enabled plugins file and plugins directory are also used to locate plugins commands